3주차 과제: 테스트 코드 작성하기

Ubermensch0608 opened the pull request:
안녕하세요. 트레이너님 3주차 과제 PR 올립니다!

테스트 코드를 작성하는 것에 있어 어떤 matcher를 사용해야 할지 등 익숙치 않은 것이 많지만, 무엇보다 어떤 테스트 코드를 작성해야할까? 가 어려운 부분인 것 같습니다. 최근 테스트 트렌드는 내부적으로 어떤 로직들이 작동하는 지 보다, 실제로 유저에게 요구사항에 맞게 잘 보이거나 동작하는가가 중요하다고는 들었습니다. 
하지만, 얼마나 디테일하게 테스트 코드를 작성해야하는 지 감이 잘 잡히지 않네요.  예를 들어, input 창을 이용하여 타이핑할 때 실제로 값이 잘 들어가 노출이 되는지, 버튼을 눌렀을 때 테스크에 추가가 되는 지, label에 알맞은 텍스트 값이 들어 갔는 지, label을 클릭했을 때 input에 포커싱이 갔는지 등을 모두 테스트하는 것이 좋을까요? 
테스트는 디테일 할 수록 좋은지, 빠르게 구현하는 방향이 좋은지 조언 부탁드리겠습니다..!

Ubermensch0608 committed:
Feat: 컴포넌트 별 test 구현

각 컴포넌트의 관심사에 맞는 test 구현했습니다.  비즈니스 로직에서 요구할 만한 사항 들을 나누어 컴포넌트 별 테스트 파일에 구현했습니다.

daadaadaah commented on src/App.test.jsx:
```diff
@@ -0,0 +1,27 @@
+import { render, screen } from '@testing-library/react';
+
+import Page from './Page';
```
App 테스트 파일에서 Page 컴포넌트를 테스트를 하고 있습니다! App 컴포넌트 테스트로 수정해주세요!

daadaadaah commented on src/App.test.jsx:
```diff
@@ -0,0 +1,27 @@
+import Page from './Page';
+
+describe('App', () => {
+  it('taskTitle의 value는 input에 그대로 노출되어야 한다.', () => {
```
BDD에 대해 한번 공부해봅시다!

[참고]
- https://johngrib.github.io/wiki/junit5-nested/#describe---context---it-%ED%8C%A8%ED%84%B4
- https://marocchino.net/2016/12/04/about-rspec/
- https://ko.javascript.info/testing-mocha#ref-888
- https://github.com/ahastudio/til/blob/main/ruby/20161206-rspec-let.md
- [Ginkgo - Go 언어 개발자를 위한 BDD 테스팅 프레임워크](https://www.youtube.com/watch?v=gfTsSBRvdqI)

Ubermensch0608 replied:
참고자료 보고 BDD에 대해서 공부해봤습니다!
프론트엔드에 제 나름대로 적용해 본 결과, 다음과 같이 정리해보았습니다.
- `describe`: 대상 -> 컴포넌트or태그
- `context`: 상황, 조건 ->  어떤 상태(prop, state)를 가진 경우,
- `it`: 행동, **결과** -> 어떤 형태로 화면에 **노출** 되는지 

BDD를 적용하면 테스트 결과를 깔끔한 문장형태로 정리해 볼 수 있다는 점이 너무 마음에 드네요!
피드백 반영하면서 BDD의 원칙을 전반적으로 적용해보려고 합니다! :)

daadaadaah commented on src/List.test.jsx:
```diff
@@ -0,0 +1,26 @@
+import List from './List';
+
+describe('List', () => {
+  it('할 일 목록이 없을 때 "할 일이 없어요!"를 화면에 노출시켜야 한다.', () => {
+    const tasks = [];
+    render(<List tasks={tasks} />);
+
+    const list = screen.getByRole('paragraph');
+
+    expect(list).toHaveTextContent('할 일이 없어요!');
+  });
+
+  it('할 일 목록이 있을 때, 화면에 노출시켜야 한다.', () => {
+    const tasks = [{
+      id: 0,
+      title: '안녕하세요',
+    }];
+    render(<List tasks={tasks} />);
+
+    const list = screen.getByRole('listitem');
+
+    expect(list).toHaveTextContent(/안녕하세요/i);
+  });
```
이렇게 케이스가 대비되는 상황에서는 [context](https://www.betterspecs.org/#contexts)를 한번 활용해보세요!

Ubermensch0608 replied:
넵! BDD의 계층을 지켜 어떤 상황, 조건이 나뉠 때는 `context`로 분기되도록 활용해보겠습니다~

daadaadaah commented on src/Input.test.jsx:
```diff
@@ -0,0 +1,48 @@
+
+describe('Input', () => {
+  it('label은 할 일 이라는 텍스트를 가지고 있어야 한다.', () => {
+    render(<Input />);
```
render 함수가 계속 중복되는거 보이시나요? 테스트 코드도 리팩토링하면 좋아요!
중복을 한번 제거해봅시다!

daadaadaah commented on src/Item.test.jsx:
```diff
@@ -10,19 +10,21 @@ test('Item', () => {
   expect(container).toHaveTextContent('뭐라도 하기');
   expect(container).toHaveTextContent('완료');
 
+  const button = getByText('완료');
```
저는 버튼의 경우, 버튼이라는 의도가 더 드러나도록 `getByRole('button', {name: /submit/i}`를 활용하는 편입니다

Ubermensch0608 replied:
대상을 잡을 때 좀 더 의도를 들어낼 수 있도록(특정할 수 있도록) 해보겠습니다!

daadaadaah commented on src/Input.test.jsx:
```diff
@@ -0,0 +1,48 @@
+    expect(input.value).toBe('누워 있기');
+  });
+
+  it('확인 버튼을 누르면 빈 value로 화면에 노출되어야 한다.', async () => {
+    const addTodo = jest.fn();
+    const changeInputValue = jest.fn();
+
+    render(<Input onClick={addTodo} onChange={changeInputValue} />);
+
+    const input = screen.getByRole('textbox');
+
+    fireEvent.change(input, { target: { value: '고양이 밥 주기' } });
+
+    const button = screen.getByRole('button', { name: '추가' });
+
+    expect(screen.getByRole('textbox').value).toBe('고양이 밥 주기');
+
+    fireEvent.click(button);
+    expect(addTodo).toHaveBeenCalledTimes(1);
+
+    fireEvent.change(input, { target: { value: '' } });
+    expect(screen.getByRole('textbox').value).toBe('');
+  });
```
테스트당 1가지 목적을 가지고 테스트하는 것이 좋아요!
지금 한가지 테스트로 너무 많은 걸 알려고 하고 있어요ㅠㅠ

Ubermensch0608 replied:
결국 원칙이 없으면 코드가 뒤죽박죽 되어버리네요..하하
BDD 원칙에 따라 리팩토링 중인데 많이 깔끔해지는 거 같습니다!

daadaadaah commented on src/App.test.jsx:
```diff
@@ -0,0 +1,27 @@
+    const addTask = jest.fn();
+    const deleteTask = jest.fn();
+
+    render(
+      <Page
+        taskTitle={taskTitle}
+        tasks={tasks}
+        onChangeTitle={changeTaskTitle}
+        onClickAddTask={addTask}
+        onClickDeleteTask={deleteTask}
+      />,
+    );
+
+    const title = screen.getByRole('textbox');
+    expect(title).toHaveValue('테스크 타이틀');
```
현재 많은 테스트에서 getByRole로 요소를 특정하고, 그 값을 확인하고 있는데, 이렇게 되면 구현에 포커스된 테스트인 것 같아요! 우리가 기대하는거는 Page 컴포넌트에 어떤 텍스트들이 보이냐이므로,        container를 활용해보셔도 좋을 것 같아요!
```js
const { container } = render~~~~;
expect(container).toHaveTextContent("기대하는 값");
```

daadaadaah reviewed (changes requested):
> 테스트 코드를 작성하는 것에 있어 어떤 matcher를 사용해야 할지 등 익숙치 않은 것이 많지만, 무엇보다 어떤 테스트 코드를 작성해야할까? 가 어려운 부분인 것 같습니다. 최근 테스트 트렌드는 내부적으로 어떤 로직들이 작동하는 지 보다, 실제로 유저에게 요구사항에 맞게 잘 보이거나 동작하는가가 중요하다고는 들었습니다.
하지만, 얼마나 디테일하게 테스트 코드를 작성해야하는 지 감이 잘 잡히지 않네요. 예를 들어, input 창을 이용하여 타이핑할 때 실제로 값이 잘 들어가 노출이 되는지, 버튼을 눌렀을 때 테스크에 추가가 되는 지, label에 알맞은 텍스트 값이 들어 갔는 지, label을 클릭했을 때 input에 포커싱이 갔는지 등을 모두 테스트하는 것이 좋을까요?
1. 테스트 코드 작성
- 맞습니다! 테스트 코드 작성에 있어 가장 선결적으로 해결해야하는 것이 "무엇"을 테스트할 것인가입니다! 제가 테스트를 어떻게 발견하는지와 도움될만한 링크 남겨놓았습니다! 한번 참고하세요!
- [https://github.com/CodeSoom/react-week3-assignment-1/pull/134#pullrequestreview-1013694078](https://github.com/CodeSoom/react-week3-assignment-1/pull/134#pullrequestreview-1013694078)
- [https://github.com/CodeSoom/react-week3-assignment-1/pull/138#pullrequestreview-1015192093](https://github.com/CodeSoom/react-week3-assignment-1/pull/138#pullrequestreview-1015192093)

2. 어떤 matcher를 사용해야하는지에 대해서 저는 아래 링크 보고 참고합니다!
- https://github.com/testing-library/jest-dom#custom-matchers

Ubermensch0608 committed:
Refactor: Input 컴포넌트 테스트 리팩토링

BDD 원칙을 생각하며 테스트 수정 구현했습니다.

daadaadaah commented on src/Input.test.jsx:
```diff
@@ -1,48 +1,55 @@
-    render(<Input />);
+describe('Input 컴포넌트', () => {
+  const handleAddText = jest.fn();
+  const handleChangeText = jest.fn();
+  const renderInput = render(
```
맥락 구분을 위해 한줄 띄어줍시다!

daadaadaah commented on src/Input.test.jsx:
```diff
@@ -1,48 +1,55 @@
+  const input = getByPlaceholderText('할 일을 입력해 주세요');
 
-    expect(input.value).not.toBe('누워 있기');
+  context('최초 렌더링 되었을 때', () => {
```
[context](https://www.betterspecs.org/#contexts)는 주로 상반되는 상황에서 사용합니다. 예를 들어, `할일이 있다면/할일이 없다면` 과 같이요!
지금 상황에서는 그냥 it문으로만 표현해주셔도 될 것 같아요!

daadaadaah commented on src/Input.test.jsx:
```diff
@@ -1,48 +1,55 @@
+    fireEvent.click(button);
 
-    expect(screen.getByRole('textbox').value).toBe('고양이 밥 주기');
+    it('handleAddText가 한 번 실행되어야 한다.', () => {
```
`~되어야 한다` 표현보다 `~된다.`라 표현을 주로 사용합니다.

daadaadaah commented on src/Input.test.jsx:
```diff
@@ -1,48 +1,55 @@
-    fireEvent.change(input, { target: { value: '누워 있기' } });
-
-    expect(input.value).toBe('누워 있기');
+    it('빈 값으로 노출되어야 한다.', () => {
```
BDD는 테스트 대상의 `행동`을 중점을 두고 테스트 코드를 작성하는 것으로, `노출된다`라는 표현도 있지만 `렌더된다`라는 표현으로 써도 괜찮을 것 같아요!ㅎㅎ

daadaadaah commented on src/Input.test.jsx:
```diff
@@ -1,48 +1,55 @@
-  it('label은 할 일 이라는 텍스트를 가지고 있어야 한다.', () => {
-    render(<Input />);
+describe('Input 컴포넌트', () => {
+  const handleAddText = jest.fn();
```
Text 보다 좀더 의도가 드러나는 표현은 없을까요?

Ubermensch0608 committed:
Refactor: BDD를 적용한 테스트 코드 작성

각 컴포넌트에 대한 테스트를 독립적인 것으로 생각하면서 테스트 코드 리팩토링 했습니다.
BDD 원칙에 따라 상반된 상황에서 분기를 나누는 등 논리적인 구조로 작성했습니다.
react testing library의best practice를 적용했습니다.

daadaadaah commented on src/Page.test.jsx:
```diff
@@ -0,0 +1,46 @@
+
+import Page from './Page';
+
+const renderPageWithNoTasks = () => {
+  const tasks = [];
+
+  render(<Page tasks={tasks} />);
+};
```
34번째 줄 테스트에서도 이 함수를 활용하기 위해서는 함수를 어떻게 만드는것이 좋을까요?

daadaadaah commented on src/List.test.jsx:
```diff
@@ -0,0 +1,30 @@
+  });
+
+  context('할 일 목록이 있을 때', () => {
+    const tasks = [
+      { id: 0, title: '첫 번째' },
+      { id: 1, title: '두 번째' },
+    ];
```
tasks 는 다른 테스트에서도 사용되는 것을 알 수 있습니다. 
이런 테스트영 데이터를 `fixtrue`라고 하는데, 저는 `fixtures` 폴더를 하나 만들어서 따로 관리하는 편입니다.
